udata_recommendations/commands.py

```diff
@@ -49,52 +71,80 @@ def get_dataset(id_or_slug):
-        dataset_obj.save()
+        log.info(f"Found {len(valid_recos)} new recommendations for dataset {dataset['id']}")
+
+        new_sources = set(target_dataset.extras.get('recommendations:sources', []))
+        new_sources.add(source)
+
+        merged_recommendations = target_dataset.extras.get('recommendations', [])
+        merged_recommendations.extend(valid_recos)
+        new_recommendations = sorted(merged_recommendations, key=lambda k: k['score'], reverse=True)
+
+        target_dataset.extras['recommendations:sources'] = list(new_sources)
+        target_dataset.extras['recommendations'] = new_recommendations
+        target_dataset.save()
     else:
-        error('No recommendation found for dataset %s' % dataset['id'])
-        return 0
-    return 1
+        log.error(f"No recommendation found for dataset {dataset['id']}")
+
+
+def process_sources(sources, should_clean):
+    for source, url in sources.items():
+        if should_clean:
+            log.info(f'Cleaning up dataset recommendations from source {source}')
+            clean_datasets_recommendations(source)
+
+        log.info(f'Fetching dataset recommendations from {url}, source {source}')
+        process_source(source, get_recommendations_data(url))
+
+
+@recommendations.command()
+@click.argument('source')
+def clean(source):
+    clean_datasets_recommendations(source)
 
 
 @recommendations.command()
-@click.option('-s', '--source', default=None,
-              help='The source URL to get recommendations from')
-@click.option('-c', '--clean', is_flag=True, default=True,
-              help='Cleanup existing recommendations')
-def datasets(source, clean):
-    if clean:
-        clean_datasets_recommendations()
-    nb_datasets = 0
-    url = source or current_app.config.get(
-        'RECOMMENDATIONS_DATASETS_SOURCE_URL')
-    if not url:
-        exit_with_error('Not source url set, aborting.')
-    log.info('Fetching dataset recommendations from %s...', url)
-    data = get_datasets_data(url)
-    for idx, dataset in enumerate(data):
-        nb_datasets += process_dataset(idx, dataset)
-    success('Dataset recommendations filled for %s dataset(s).' % nb_datasets)
+@click.option('-u', '--url', default=None, help='The URL to get recommendations from')
+@click.option('-s', '--source', default=None, help='The source name associated with this URL')
+@click.option('--clean', is_flag=True, help='Clean recommendations before fetching recommendations')
+def add(url, source, clean):
+    use_config = url is None and source is None
+
+    if use_config:
+        sources = current_app.config.get('RECOMMENDATIONS_SOURCES')
+    else:
+        if url is None or source is None:
+            exit_with_error('You should specify a source and a URL')
+
+        sources = {source: url}
+
+    try:
+        process_sources(sources, clean)
+    except jsonschema.exceptions.ValidationError as e:
+        exit_with_error(f"Fetched data is invalid {str(e)}")
```

Comment:
I'm wondering if this try/except dance is needed as it could bubble up and just crash the command

Reply:
True. Not really a problem though, as you wish.